plugins/integrations/prometheus/src/org/apache/cloudstack/metrics/PrometheusExporterImpl.java

```diff
@@ -0,0 +1,612 @@
+        metricsList.add(new ItemVlan(zoneName, zoneUuid, ALLOCATED, dcDao.countZoneVlans(dcId, true)));
+        metricsList.add(new ItemVlan(zoneName, zoneUuid, TOTAL, dcDao.countZoneVlans(dcId, false)));
+    }
+
+    private void addDomainLimits(final List<Item> metricsList) {
+        Long totalCpuLimit = 0L;
+        Long totalMemoryLimit = 0L;
+
+        for (final DomainJoinVO domain: domainDao.listAll()) {
+            if (domain == null || domain.getLevel() != 1) {
+                continue;
+            }
+            long cpuLimit = ApiDBUtils.findCorrectResourceLimitForDomain(domain.getCpuLimit(), false,
+                    Resource.ResourceType.cpu, domain.getId());
+            if (cpuLimit > 0) {
+                totalCpuLimit += cpuLimit;
+            }
+
+            long memoryLimit = ApiDBUtils.findCorrectResourceLimitForDomain(domain.getMemoryLimit(), false,
+                    Resource.ResourceType.memory, domain.getId());
+            if (memoryLimit > 0) {
+                totalMemoryLimit += memoryLimit;
+            }
+        }
+        metricsList.add(new ItemDomainLimitCpu(totalCpuLimit));
+        metricsList.add(new ItemDomainLimitMemory(totalMemoryLimit));
+    }
+
+    @Override
+    public void updateMetrics() {
+        final List<Item> latestMetricsItems = new ArrayList<Item>();
+        try {
+            for (final DataCenterVO dc : dcDao.listAll()) {
+                final String zoneName = dc.getName();
+                final String zoneUuid = dc.getUuid();
+                alertManager.recalculateCapacity();
+                addHostMetrics(latestMetricsItems, dc.getId(), zoneName, zoneUuid);
+                addVMMetrics(latestMetricsItems, dc.getId(), zoneName, zoneUuid);
+                addVolumeMetrics(latestMetricsItems, dc.getId(), zoneName, zoneUuid);
+                addStorageMetrics(latestMetricsItems, dc.getId(), zoneName, zoneUuid);
+                addIpAddressMetrics(latestMetricsItems, dc.getId(), zoneName, zoneUuid);
+                addVlanMetrics(latestMetricsItems, dc.getId(), zoneName, zoneUuid);
+            }
+            addDomainLimits(latestMetricsItems);
+        } catch (Exception e) {
+            LOG.warn("Getting metrics failed ", e);
+        }
+        metricsItems = latestMetricsItems;
+    }
+
+    @Override
+    public String getMetrics() {
+        final StringBuilder stringBuilder = new StringBuilder();
+        stringBuilder.append("# Cloudstack Prometheus Metrics\n");
+        for (final Item item : metricsItems) {
+            stringBuilder.append(item.toMetricsString()).append("\n");
+        }
+        return stringBuilder.toString();
+    }
+
+    private abstract class Item {
+        String name;
+
+        public Item(final String nm) {
+            name = nm;
+        }
+
+        public abstract String toMetricsString();
+    }
+
+    class ItemVM extends Item {
```

Comment:
a lot of doublures in fields i.e. zoneName and zoneUuid or even zoneFilter. maybe a different hierarchy is an idea.

Reply:
What approach should we take?